main.cpp

```diff
@@ -130,81 +130,81 @@ You are going to write 10 UDTs in project3.
  */
 
 /*
-Thing 5)
+Thing 5) Harmonic set (generates a set of frequencies from two input frequencies)
 5 properties:
-    1)
-    2)
-    3)
-    4)
-    5)
+    1) Bass Frequency
+    2) Generator Frequency 
+    3) Minimum output frequency
+    4) maximum output frequency
+    5) generator rule # (would select among a few pre-coded rules)
 3 things it can do:
-    1)
-    2)
-    3)
+    1) calculate and output a set of frequencies
+    2) calculate statistics about the frequencies (how harmonically coherent they are, etc)
+    3) filter the frequency set according to certain rules
  */
 
 /*
-Thing 6)
+Thing 6) Rhythmic Rules
 5 properties:
-    1)
-    2)
-    3)
-    4)
-    5)
+    1) density / note gap %
+    2) "convolution rate" (how complicated the rhythms are)
+    3) tempo
+    4) rate of variation (how far it randomly strays from the selected settings)
+    5) meter / pulse structure (this would have to be a UDT also)
 3 things it can do:
-    1)
-    2)
-    3)
+    1) output a rhythm that controls the pattern generator
+    2) modify the rhythmic properties according to a control signal from the UI
+    3) modify the meter and tempo according to UI control
  */
 
 /*
-Thing 7)
+Thing 7) Pattern Generator
 5 properties:
-    1)
-    2)
-    3)
-    4)
-    5)
+    1) "rootedness" (how often it plays the generator frequencies)
+    2) repetition rate (how likely to repeat notes)
+    3) step size (how far away in the set it will pick the next note)
+    4) variability rate
+    5) fractal dimension (would be used in the pattern calculation)
 3 things it can do:
-    1)
-    2)
-    3)
+    1) receives inpulses from the rhythm generator
+    2) calculates next note based on current harmonic set
+    3) outputs control signals to the synth
  */
 
 /*
-Thing 8)
+Thing 8) Synthesizer
 5 properties:
-    1)
-    2)
-    3)
-    4)
-    5)
+    1) wave shape (int, selects between pre-coded waveshapes)
+    2) amplitude
+    3) envelope
+    4) max polyphony
+    5) distortion
```

Comment:
some of these are objects, not primitives!   You need to pick things that can be represented with `int float double bool char std::string`
- [ ] revise